app/code/Magento/Catalog/Pricing/Render/FinalPriceBox.php

```diff
@@ -64,26 +64,36 @@ public function __construct(
      */
     protected function _toHtml()
     {
-        if (!$this->salableResolver->isSalable($this->getSaleableItem())) {
-            return '';
-        }
-
         $result = parent::_toHtml();
-        //Renders MSRP in case it is enabled
-        if ($this->isMsrpPriceApplicable()) {
-            /** @var BasePriceBox $msrpBlock */
-            $msrpBlock = $this->rendererPool->createPriceRender(
-                MsrpPrice::PRICE_CODE,
-                $this->getSaleableItem(),
-                [
-                    'real_price_html' => $result,
-                    'zone' => $this->getZone(),
-                ]
-            );
-            $result = $msrpBlock->toHtml();
-        }
 
-        return $this->wrapResult($result);
+        if(!$result) {
+            $result = BasePriceBox::_toHtml();
+            try {
+                /** @var MsrpPrice $msrpPriceType */
+                $msrpPriceType = $this->getSaleableItem()->getPriceInfo()->getPrice('msrp_price');
+            } catch (\InvalidArgumentException $e) {
+                $this->_logger->critical($e);
+                return $this->wrapResult($result);
```

Comment:
I'd refactor everything related to `$msrpPriceType` into the try-catch in which `$msrpPriceType` is defined, this way the return logic (`$this->wrapResult($result)`) is not duplicated.

I can understand if this was done to prevent too much indentation by using an early return. In this case I'd argue that it might be better to refactor it into a separate method responsible for returning either the HTML for the MSRP price rendering or an empty string in case of the exception for example, minimising the overal indentation.